### PR TITLE
Unified Version Proposal: NO VERSIONING

### PR DESCRIPTION
**Pull request by ZmnSCPxj-jr**
What is better than unifying versions?  **NOT HAVING TO IMPLEMENT VERSIONS AT ALL**.

The advantages are:

* No more versions.
  * Less code: no need to check versions.
  * Less roundtrips: no questioning of versions and selecting the version.
* Much less amibiguity in specifications:
  * Do we have separate "For implementation", "Stable" etc. states for each VERSION or for each LSPS?

In particular, we SHOULD in fact separate LSPS2 from LSPS4 (***YES*** @SeverinAlexB ***I AM CHANGING MY MIND ABOUT THIS***).  The reason is that LSPS2 has  *fundamentally different flow* from LSPS4:

* LSPS2:
  * Get quotes on how much JIT channels cost.
  * Buy an SCID.
  * Issue invoice
  * Receive payment and get JIT channel opened.
* LSPS4:
  * Get the permanent SCID
  * Issue invoice.
  * Receive payment notification. If it fits in current inbound liquidity, use it and end the flow.  Otherwise continue...
  * Get quotes on how much JIT channels cost.
  * Buy a JIT channel open.
  * Reeive payment and get JIT channel opened.

Thus, we ***MUST*** in fact put them as separate specifications completely, because they the fundamental differences in flow.

Even with no-version, we can still upgrade individual specifications:

* We can add optional parameters to API requests:
  * If not specified, run back-compatibly with previous versions.
  * If specified, new version.  If LSP does not recognize, use standard "parameter error" `-32602` from JSON-RPC 2.0 so that client knows LSP does not support new version.
* We can add new fields in API responses or LSP notifications:
  * LSPS0 specifically requires that clients ignore unrecognized fields in responses and notifications.
  * If client knows it but it does not appear, then LSP is on older version.

Basically:

* If the changes are big enough that they cannot fit in the above back-compatibility, then we really MUST put them in a new LSPS.
* Otherwise, the changes are small enough that they do not need a separate LSPS but can fit in the same flow as an existing LSPS.

Putting LSPS2 and LSPS4 specifically in separate LSPS numbers also makes it much easier to reason about the LSPS states "For Implementation" "Stable" and "Deprecated".  Until LSPS4 is stable, we should not deprecate LSPS2.  If we move LSPS4 as LSPS2v2 then we need separate states for *each version* of an LSPS.

No-version FTW.

### CODE REVIEW

**Comment by kingonly** [comment]
Ack

**Comment by SeverinAlexB** [comment]
So this proposal removes versioning from LSPS2 and keeps it at LSPS1? Every LSPS does it's own thing?

I am happy to go the "every breaking change is its own LSPS" route which is very similar to my "we run all versions". But having consistency in the approach on how to handle breaking changes is a requirement.

At the end, you always have versions. Implicit or explicit. A section in the LSPS0 on the approach taken is definitely needed though.

**Comment by ErikDeSmedt** [comment]
CONCEPT_ACK

To merge this PR we still need to remove versioning from LSPS1 and possibly other places.

**Comment by ZmnSCPxj-jr** [comment]
As noted in Telegram group: We now need to define the -32602 error `code` (invalid parameters) in detail.  **Rationale** Suppose we have a future version of an RPC method that adds the optional parameter `v2param`,  Then, suppose have an even later version of that RPC method that adds the optional parameter `v3param`.  If the client provides both `v2param` and `v3param`, and the LSP returns with a -32602 invalid parameter error (meaning it probably does not support the latest version), how does the client know if the LSP supports `v2param` but not `v3param`, or if the LSP does not support `v2param` nor `v3param`?  Thus, we need to add an `error.data` object with a field `error.data.unrecognized`, a JSON array of strings representing the set of unrecognized parameters.

**Comment by ZmnSCPxj-jr** [comment]
updated: Describe `unrecognized` field of data of -32602 "Invalid params" errors, remove versioning in LSPS1.

**Comment by ZmnSCPxj-jr** [comment]
>Or do we want to keep it to signal spec changes and just keep versioning out of the actual protocol?

I think keep it to signal spec changes.

File: LSPS1/README.md [hunk 1]
```diff
@@ -122,6 +119,8 @@ The client MUST call `lsps1.info` first.
 
 Every `min/max` options pair MUST ensure that `min <= max`.
 
+An LSP, and newer versions of this LSPS, MUST NOT add new fields to the `options` dictionary.
+Future versions of this LSPS, and extensions to this LSPS, MAY add new entire fields in the result, such as adding a new `v2_options` dictionary, which only apply if the client speifically requests for using those options.
```
Comment (SeverinAlexB) [inline LSPS1/README.md:123]: I don't know if I understand this correctly so let me know if I made any logical mistakes:

- In LSPS1, we are not allowed to add an optional option. What is the reason behind this?
- In future LSPSx, we should be able to change whatever we want. Why are we not allowed to completely redo `options`?

**Review by SeverinAlexB — COMMENTED** [review]
Apart from my one comment, this PR looks good to me.

File: LSPS1/README.md (hunk 1 above)
Comment (ZmnSCPxj-jr) [reply LSPS1/README.md:123]: >In LSPS1, we are not allowed to add an optional option. What is the reason behind this?

Because the additional option may imply that the characteristics of the channel may be different from what a pre-this-new-option client expects.  This is why a future revision of LSPS1 that wants to add a new option needs to put those in a new `v2_options` field, and which the client must specifically opt-into so that the LSP knows that the client understands the new options.

>In future LSPSx, we should be able to change whatever we wan

True.  That is why the text says "this LSPS" and not "any future LSPS".

**Comment by ZmnSCPxj-jr** [comment]
Updated: Expand LSPS0 "LSPS Extension and Versioning" section to include client and LSP MUSTs, remove remaining reference in LSPS0 re: versioning API calls in other LSPS.

File: LSPS1/README.md (hunk 1 above)
Comment (SeverinAlexB) [reply LSPS1/README.md:123]: > True. That is why the text says "this LSPS" and not "any future LSPS".

I have trouble understanding what is "Future versions of this LSPS". It is clear now because you elaborated on it but it is hard to get just from reading the PR.
Problem: From a language POV, even a non-breaking change is a new version.
Can we define standardize language around this so it gets very clear what the spec means?
Borrowing definitions from [SemVer](https://semver.org/):
- "Major" version when you make incompatible API changes.
- "Minor" version when you add functionality in a backward compatible manner.

What do you think?

>Because the additional option may imply that the characteristics of the channel may be different from what a pre-this-new-option client expects. This is why a future revision of LSPS1 that wants to add a new option needs to put those in a new v2_options field, and which the client must specifically opt-into so that the LSP knows that the client understands the new options.

Is this necessary? Is there no way to add an additional option if the new LSPS1 guarantees to use backward compatible default options from the previous (minor) version?

Comment (ZmnSCPxj-jr) [reply LSPS1/README.md:123]: > Is there no way to add an additional option if the new LSPS1 guarantees to use backward compatible default options from the previous (minor) version?

What happens if the LSP gives a non-default option that is defined in a later version?

**Comment by ZmnSCPxj-jr** [comment]
Updated: Explicitly state in LSPS0 that new LSPS must be back-compatible, remove restriction on `options` in LSPS1.

File: LSPS1/README.md (hunk 1 above)
Comment (ZmnSCPxj-jr) [reply LSPS1/README.md]: Resolved in 2023-11-16: remove the restriction on `options`.

**Comment by ZmnSCPxj-jr** [comment]
Updated: Add a few more details in what future LSPS is allowed to do, and what client must do.

File: LSPS0/README.md [hunk 2]
```diff
@@ -391,6 +379,66 @@ SHOULD indicate this as an "unrecognized" error to the user.
...
+{
+  "jsonrpc": "2.0",
+  "method": "example.method_name",
+  "params": {
+    "v1_parameter": "value1",
+    "v2_parameter": "value2"
+  },
+  "id": "42"
+}
+```
```
Comment (ErikDeSmedt) [inline LSPS0/README.md:406]: Can we rephrase to speak in terms of `features` and not in terms of `versions`?

It is somewhat confusing to use the wording versions here and explain in the last section that we don't want to support a versioning scheme.



```JSON
{
  "jsonrpc": "2.0",
  "method": "example.method_name",
  "params": {
    "v1_feature": "value1",
    "v2_feature": "value2"
  },
  "id": "42"
}
```

File: LSPS0/README.md [hunk 3]
```diff
@@ -391,6 +379,66 @@ SHOULD indicate this as an "unrecognized" error to the user.
...
+> an even newer parameter `v3_parameter` for that method.
+> If a client sends a request that includes `v2_parameter` *and*
+> `v3_parameter`, and the LSP does not yet support the latest
+> version of the LSPS, the LSP would return an "Invalid params"
+> -32602 error.
+> However, without the `unrecognized` field in the `data` object,
+> the client cannot know if the LSP does not support `v2_parameter`
+> *and* `v3_parameter`, or supports `v2_parameter` but not
+> `v3_parameter`.
+>
```
Comment (ErikDeSmedt) [inline LSPS0/README.md:440]: This text seems to imply that changes in the spec must be implemented sequentially. (Version numbers)

I don't think this has to be true. 
1. Suppose we update LSPS2 to include splicing. (`v2` or `feature_splicing`)
2. Later we'll add the functionality to purchase channels on behalf of others (`v3` or `feature_buy_peer`)

An LSP-server might struggle to implement `feature_splicing` because they are waiting for their node implementation to support it. But they might be able to implement `feature_buy_peer` as soon as the spec is completed.

In terms of features you can implement `feature_buy_peer` before `feature_splicing`.
In terms of features it is weird to implement `v3` before you implement `v2`.

Yeah, I like the feature-wording much better 🤓

File: LSPS0/README.md [hunk 4]
```diff
@@ -525,6 +573,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+
+* Ignore unrecognized `result` fields from the result of a
+  client-callable API `method`.
+  * Recursively ignore unrecognized fields from any dictionaries
+    in a recognized `result` field, to any depth.
+* Treat as a generic error, and log as unusual, any unrecognized
+  error `code` from an `error` result for a client-callable
+  API `method`.
+* Ignore, but log as unusual, unrecognized LSP-initiated
+  notification `method`s.
```
Comment (ErikDeSmedt) [inline LSPS0/README.md:629]: NIT

`MUST` is fierce wording here. Most users don't even realize their mobile app has logs. 
I wouldn't consider a client non-compliant because of a poor logging implementation.

File: LSPS0/README.md [hunk 5]
```diff
@@ -525,6 +573,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+
+LSPs MUST:
```
Comment (ErikDeSmedt) [inline LSPS0/README.md:631]: NIT: Could you use LSP-servers here.

File: LSPS0/README.md (hunk 3 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:440]: ok

File: LSPS0/README.md (hunk 5 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:631]: Hmmm why "LSP-servers"?

File: LSPS0/README.md (hunk 4 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:629]: So still MUST ignore but remove the bit about logging?

File: LSPS0/README.md (hunk 5 above)
Comment (ErikDeSmedt) [reply LSPS0/README.md:631]: I think LSPS (LSP-specifications) vs LSPs (Lightning Service Providers) is confusing here.

File: LSPS0/README.md (hunk 4 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:629]: Put the logging bits as `SHOULD`.

File: LSPS0/README.md (hunk 5 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:631]: Reworded to `An LSP MUST`

**Comment by ZmnSCPxj-jr** [comment]
Updated: rebase to latest, avoid "version" and use "revision" (for specifications) or "feature" (for params/retvals) instead, use "An LSP MUST" instead of "LSPs MUST", move client MUST log to SHOULD log.

File: LSPS0/README.md (hunk 3 above)
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md]: done

**Review by JssDWt — APPROVED** [review]
LGTM

**Review by SeverinAlexB — APPROVED** [review]
Let's do it

File: LSPS0/README.md [hunk 6]
```diff
@@ -525,6 +571,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+
+Later revisions of an LSPS MAY, if a new need arises:
+
+* Add new optional `params` fields for a client-callable API `method`.
+  * If the optional parameter is not specified, the `method` MUST
+    act the same as in previous versions before the parameter
+    existed.
+* Add new required or optional `result` fields for a client-callable
+  API `method`.
+  * Provided that the client is free to ignore the added fields and
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:595]: not modified due to "no harm" anyway.

File: LSPS0/README.md [hunk 7]
```diff
@@ -525,6 +571,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+
+If a client wants to use a newer version of an LSPS, it can look
+for some new `result` field, and if that does not exist, it knows
+the LSP supports an older version of the LSPS.
+The client can provide any new optional parameters it wants to
+use, and if the LSP responds with a -32602 Invalid params error,
+that error includes an `unrecognized` field, described above, that
+contains the parameters that the LSP does not supoort.
+The client can use a new client-callable API `method`, and if the
+LSP responds with a -32601 Method not found error, knows that the
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:652]: done

File: LSPS0/README.md [hunk 8]
```diff
@@ -525,6 +571,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+If the client supports an older version of an LSPS, it simply does
+not provide any new optional parameters to existing client-callable
+API `method`s, or any new client-callable API `method`s.
+Then the interface would act the same as in the older version.
+
+If a client wants to use a newer version of an LSPS, it can look
+for some new `result` field, and if that does not exist, it knows
+the LSP supports an older version of the LSPS.
+The client can provide any new optional parameters it wants to
+use, and if the LSP responds with a -32602 Invalid params error,
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:648]: done.

File: LSPS0/README.md [hunk 9]
```diff
@@ -525,6 +571,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+    act the same as in previous versions before the parameter
+    existed.
+* Add new required or optional `result` fields for a client-callable
+  API `method`.
+  * Provided that the client is free to ignore the added fields and
+    that acceptance of a new `result` field is signalled by using a
+    new optional `params` field of some client-callable API `method`,
+    or by calling a new client-callable API `method`, or by some other
+    method, and that a lack of signal of acceptance of the new field
+    results in behaving the same as in older versions.
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:600]: done

File: LSPS0/README.md [hunk 10]
```diff
@@ -525,6 +571,91 @@ payload, indicating it supports LSPS1 and LSPS3 (in addition to LSPS0):
...
+
+* Respond with a -32602 "Invalid params" error, described in a
+  previous section with `unrecognized` field in the `data`
+  dictionary, if it receives a client-callable API `method` it
+  recognized, but with an unrecognized parameter.
+* Respond with a -32601 "Method not found" error, if it receives a
+  request to call a client-callable API `method` that it does not
+  recognize.
+
+If the client supports an older version of an LSPS, it simply does
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS0/README.md:639]: done

**Comment by ZmnSCPxj-jr** [comment]
Updated: updated as per nits from @tnull

**Comment by SeverinAlexB** [comment]
Merged. Decision made in the meeting of the 7th of December 2023.
